src/librbd/cache/pwl/ssd/WriteLog.cc

```diff
@@ -124,18 +124,33 @@ bool WriteLog<I>::initialize_pool(Context *on_finish,
                                nullptr, nullptr, nullptr);
     int r = bdev->open(this->m_log_pool_name);
     if (r < 0) {
+      lderr(m_image_ctx.cct) << "fail to open block device." << dendl;
       delete bdev;
-      on_finish->complete(-1);
+      on_finish->complete(r);
+      return false;
+    }
+    if (this->m_log_pool_size != bdev->get_size()) {
+      lderr(m_image_ctx.cct) << "rbd_persistent_cache_size is not align to "
+                             << "bdev block size: " << bdev->get_block_size()
+                             << ". Please set an aligned cache size in ssd"
+                             << " mode. The closest aligned size is "
+                             << bdev->get_size() << dendl;
+      bdev->close();
+      delete bdev;
+      on_finish->complete(-EINVAL);
```

Comment:
I think the same checks should apply to an already existing cache file.  So I would move this to a separate method, to be called here and before `load_existing_entries()`.

Reply:
> I think the same checks should apply to an already existing cache file. So I would move this to a separate method, to be called here and before `load_existing_entries()`.

yes, it‘s been repeated many times, but https://github.com/ceph/ceph/pull/42046/ also have this code block. We can make it as separate method later.

Reply:
This line also need,  can we do it later?
https://github.com/ceph/ceph/blob/3111ba871f58996c1564c68d53f6c50cee922a01/src/librbd/cache/pwl/ssd/WriteLog.cc#L234

Reply:
Sorry for the delay!  Here is what I had in mind:

https://github.com/idryomov/ceph/commit/1fc722bc218acfe8a52ad7be2c06347bddb42bbe
https://github.com/idryomov/ceph/commit/58408b5f5594af3fb36ec230f76864e3d0fc5462

(instead of "librbd/cache/pwl/ssd: fix aio_write() assert due to not aligned cache size" here).

If it looks OK, I'll make the replacement.

Reply:
Never mind. Thank you for optimizing the code yourself. These two commite looks OK. Just a small suggestion about errors output. If I am a user, I want output to tell me how to fix when an error occurs, please see my comments in https://github.com/idryomov/ceph/commit/58408b5f5594af3fb36ec230f76864e3d0fc5462.

Reply:
Great, pushed!